Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -98,30 +98,105 @@ int VertexLoaderX64::ReadVertex(OpArg data, u64 attribute, int format, int count
 
 	int elem_size = 1 << (format / 2);
 	int load_bytes = elem_size * count_in;
-	if (load_bytes > 8)
-		MOVDQU(coords, data);
-	else if (load_bytes > 4)
-		MOVQ_xmm(coords, data);
-	else
-		MOVD_xmm(coords, data);
+	OpArg dest = MDisp(dst_reg, m_dst_ofs);
+
+	native_format->components = count_out;
+	native_format->enable = true;
+	native_format->offset = m_dst_ofs;
+	native_format->type = VAR_FLOAT;
+	native_format->integer = false;
 
-	PSHUFB(coords, M(&shuffle_lut[format][count_in - 1]));
+	m_dst_ofs += sizeof(float) * count_out;
 
-	if (format != FORMAT_FLOAT)
+	if (attribute == DIRECT)
+		m_src_ofs += load_bytes;
+
+	// Not very common but optimize it anyway.
+	if (format == FORMAT_FLOAT && count_in == 1)
 	{
-		// Sign extend
+		LoadAndSwap(32, scratch3, data);
+		MOV(32, dest, R(scratch3));
+		return load_bytes;
+	}
+
+	if (cpu_info.bSSSE3)
+	{
+		if (load_bytes > 8)
+			MOVDQU(coords, data);
+		else if (load_bytes > 4)
+			MOVQ_xmm(coords, data);
+		else
+			MOVD_xmm(coords, data);
+
+		PSHUFB(coords, M(&shuffle_lut[format][count_in - 1]));
+
+		// Sign-extend.
 		if (format == FORMAT_BYTE)
 			PSRAD(coords, 24);
 		if (format == FORMAT_SHORT)
 			PSRAD(coords, 16);
+	}
+	else
+	{
+		switch (format)
+		{
+		case FORMAT_UBYTE:
+		case FORMAT_BYTE:
+			MOVD_xmm(coords, data);
+			PXOR(XMM1, R(XMM1));
+			PUNPCKLBW(coords, R(XMM1));
+			if (format != FORMAT_BYTE || count_in > 1)
+				PUNPCKLWD(coords, R(XMM1));
+			if (format == FORMAT_BYTE)
+			{
+				PSLLD(coords, 24);
+				PSRAD(coords, 24);
+			}
+			break;
+		case FORMAT_USHORT:
+		case FORMAT_SHORT:
+			LoadAndSwap(64, scratch3, data);
+			MOVQ_xmm(coords, R(scratch3));
+			PSHUFLW(coords, R(coords), 0xE7); // ABCD EFGH -> ABCD EFGE
+			if (count_in == 3)
+			{
+				PUNPCKLQDQ(coords, R(coords));
+				PSHUFHW(coords, R(coords), 0xE5); // ABCD EFGH -> ABCC EFGH
+			}
+			PSLLD(coords, 16);
+			if (format == FORMAT_SHORT)
+				PSRAD(coords, 16);
+			else
+				PSRLD(coords, 16);
+			break;
+		case FORMAT_FLOAT:
+			// count_in == 1 is handled above.
+			LoadAndSwap(64, scratch3, data);
+			MOVQ_xmm(coords, R(scratch3));
```

Comment:
We don't really need to transition from general registers to XMM registers just to do the equivalent of a `ROL(R(scratch3), Imm8(32))`.